### DISCUSSION TITLE
On Hierarchy.Push() Need a little coaching.

### DISCUSSION THREAD

**Discussion by laultman** [StereoKit/StereoKit#849] [Q&A]
Never really used the Hierarchy.Push/Pop. I think that I might have a use case for it but so far baffled by how it works. I've read the docs a dozen times, and it all seems to make sense, but it doesn't work as I expected it to. So I have a collection of UI.HandleBegin "things" that each enclose meshes and models that I am drawing. They are all independent since I have no Hierarchy defined. I need to be able to move the whole collection to another location on startup to correct for where the user is and where he is looking when it first starts the app. Ideally, I want to move the scene to a known location on start up.
I thought that "wrapping" them all in a Hierarchy would allow me to adjust their pose positions without changing them individually. Am I misunderstanding the use of the Hierarchy?

**Comment by maluoi** [comment]
`Hierarchy.Push/Pop` allows you to have Object Hierarchies without a scene graph! Any draw call made within a Hierarchy will use the Hierarchy's transform as a parent.

It's also good to note that some UI Begin/End pairs also contain their own calls to `Hierarchy.Push/Pop`, such as `UI.WindowBegin/End`, and `UI.HandleBegin/End`. Basically, the Window pose gets pushed onto the Hierarchy, and is then the parent for all UI elements that get drawn to the window.

For example, in Unity you might have a GameObject hierarchy like this:
```
Player
|-Head
  |-Sunglasses
|-Hand
  |-Sword
```

To recreate this using Hierarchy.Push/Pop with StereoKit:
```csharp
Hierarchy.Push(playerTransform);
    Hierarchy.Push(headTransform);
        Hierarchy.Push(sunglassesTransform);
            sunglassModel.Draw(Matrix.Identity);
        Hierarchy.Pop();
    Hierarchy.Pop();
    Hierarchy.Push(handTransform);
        Hierarchy.Push(swordransform);
        Hierarchy.Pop();
    Hierarchy.Pop();
Hierarchy.Pop();
```

**Comment by laultman** [comment]
@maluoi Thanks for the response. I went down the path of totally unstructured "objects", for lack of a better word. My platform maintains an in-memory graph of everything it draws. I purposefully keep the graph very shallow trying to avoid the Unity style hierarchy. I think, at least to me, the SK immediate render/draw (whatever) is one of its most attractive and powerful features. It does come with trade-offs, SK draws what I want, where I want it - immediately! It is entirely my job to tell it what to draw and when.
I was hoping that I could short-circuit the graph by somehow putting everything is a big box and move the box. My platform design pattern does not cater to that concept. In fact, it would put me right back to the Unity style, I am trying to avoid.
I cater to mixed-reality, industrial workspaces. The industrial shopfloor is a chaotic place, it's like watching ants in an ant hill. The business graph lives outside the mixed reality "application" where any particular HoloLens may be executing seemingly random tasks that are not themselves interconnected at the device level.
In a large industrial plant, there are hundreds of tasks running concurrently by hundreds or thousands of individual workers. IoT is feeding the system along with all the managers, and individuals working on their job assignments. Any one HoloLens needs/must be able to pick up a task at any point is real-time and deliver content to the wearer. I worked long and hard to make Unity work in that environment and I did. The problem was then cost, maintenance, and runtime support. Customers balked. With SK and the immediate render, I've shifted the burden back out to the plant management system that has lots of horsepower. The HoloLens is an interface to the larger plant operation system. It has content pushed to it and it requests content based on user requirements.

**Comment by laultman** [comment]
@maluoi by the way, on the moving my box and all its content problem. I went back and redesigned the graph somewhat so that it adjusts the "roots" of things. By doing this, on the next frame draw the SK simply has a different Pose or position/rotation and is none the wiser.
